evm_arithmetization/src/prover.rs

```diff
@@ -522,6 +523,92 @@ fn build_segment_data<F: RichField>(
     }
 }
 
+pub struct SegmentDataIterator {
+    pub partial_next_data: Option<GenerationSegmentData>,
+    pub inputs: GenerationInputs,
+    pub max_cpu_len_log: Option<usize>,
+    pub nb_segments: usize,
+}
+
+type F = GoldilocksField;
+impl Iterator for SegmentDataIterator {
+    type Item = (GenerationInputs, GenerationSegmentData);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let cur_and_next_data = generate_next_segment::<F>(
+            self.max_cpu_len_log,
+            &self.inputs,
+            self.partial_next_data.clone(),
+        );
+
+        if cur_and_next_data.is_some() {
+            let (data, next_data) = cur_and_next_data.expect(
+                "Data cannot be
+`None`",
+            );
+            self.nb_segments += 1;
+            self.partial_next_data = next_data;
+            Some((self.inputs.clone(), data))
+        } else {
+            None
+        }
+    }
+}
+
+/// Returns the data for the current segment, as well as the data -- except
+/// registers_after -- for the next segment.
+pub(crate) fn generate_next_segment<F: RichField>(
+    max_cpu_len_log: Option<usize>,
+    inputs: &GenerationInputs,
+    partial_segment_data: Option<GenerationSegmentData>,
+) -> Option<(GenerationSegmentData, Option<GenerationSegmentData>)> {
+    let mut interpreter = Interpreter::<F>::new_with_generation_inputs(
+        KERNEL.global_labels["init"],
+        vec![],
+        inputs,
+        max_cpu_len_log,
+    );
+
+    // Get the (partial) current segment data, if it is provided. Otherwise,
+    // initialize it.
+    let mut segment_data = if let Some(partial) = partial_segment_data {
+        if partial.registers_after.program_counter == KERNEL.global_labels["halt"] {
+            return None;
+        }
+        interpreter
+            .get_mut_generation_state()
+            .set_segment_data(&partial);
+        interpreter.generation_state.memory = partial.memory.clone();
```

Comment:
I don't think the memory is trimmed here, is it? See #260 and #264

Reply:
Isn't it? `partial.memory` contains the value returned by `set_registers_and_run` (which in turn runs `run`, thus returning the trimmed `get_active_memory`). Or am I missing something?

Reply:
Yeah right, it's already trimmed!